flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/AbstractRocksDBState.java

```diff
@@ -223,4 +223,27 @@ public StateIncrementalVisitor<K, N, V> getStateIncrementalVisitor(
         throw new UnsupportedOperationException(
                 "Global state entry iterator is unsupported for RocksDb backend");
     }
+
+    /**
+     * Similar to decimal addition, add 1 to the last digit to calculate the upper bound.
+     *
+     * @param prefix the starting prefix for seek.
+     * @return end prefix for seek.
+     */
+    protected final byte[] calculateUpperBound(byte[] prefix) {
+        byte[] upperBound = new byte[prefix.length];
+        System.arraycopy(prefix, 0, upperBound, 0, prefix.length);
+        boolean overFlow = true;
+        for (int i = prefix.length - 1; i >= 0; i--) {
+            int unsignedValue = prefix[i] & 0xff;
+            int result = unsignedValue + 1;
+            upperBound[i] = (byte) (result & 0xff);
+            if (result >> 8 == 0) {
+                overFlow = false;
+                break;
+            }
+        }
+        Preconditions.checkArgument(!overFlow, "The upper boundary overflows.");
```

Comment:
If the byte array overflowed, we can append another byte in the array. We should not throw exception on mapState#clear.
Moreover, this PR lacks a test for the overflow case.

Reply:
There should be no overflow currently, because the prefix always starts with `KeyGroup`, which means that the first bit is always 0. I added this judgment mainly to prevent the `KeyGroup` from changing in the future.

Reply:
I think your implementation might not be correct. The next bytes of `0x01FF` should not be `0x0200` but was `0x01FF0`. Moreover, why the last byte would not be overflowed? (from `128` to `-127`).

Reply:
In RocksDB, `seek` is matched by the key prefix, and `0x01FF0` should be matched by the prefix of `0x01FF`, not as the ending prefix.

Since the key is regarded as an `unsigned` number in RocksDB, the next prefix for 0x7F is 0xFF, and we always use `KeyGroup` as the highest prefix of the Key, so the highest bit of the `Key` must be 0.

Reply:
According to [the current implementation](https://github.com/apache/flink/blob/1c40dc15fbc29e8e5e514565109a3fb05b47a83f/flink-runtime/src/main/java/org/apache/flink/runtime/state/CompositeKeySerializationUtils.java#L157-L159) to generate the prefix bytes, we could only have one byte to hold the array if the max key group number is not larger than 128. In this case, we cannot ensure the highest bit of the key is zero.

Reply:
Since we always write the `KeyGroup` into the prefix array first, the 0th index of the array is the highest byte, and we should add it from the highest index of the array to the 0th index.

Reply:
You're right, I missed the case where the `KeyGroup` only uses 1 byte. I'll add a unit test to cover this situation later.

Reply:
@Myasuka Sorry, I rethought this question, when the `KeyGroup` is 128, the serialized result is `0x80`, and there will be no overflow after adding 1 to the unsigned number.

Reply:
@liming30 I'm afraid that you did not get to know what is the upper limit of a single byte. `0x80` can only mean 128 if we have `0x0080` with two bytes. For the single byte, `0x80` means `-128`.

The reason why [clipDBWithKeyGroupRange](https://github.com/apache/flink/blob/master/flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBIncrementalCheckpointUtils.java#L128) would not get overflow is that we cannot change the max parallelism during restoring. Thus, we would make `currentKeyGroupRange.getEndKeyGroup() > targetKeyGroupRange.getEndKeyGroup()` as the previous max parallelism should be smaller than 128.

Reply:
Sorry, I still haven't figured out when the `KeyGroup` serialized binary array will be `0xff`or `0xffff`. If the serialization result is not `0xff` or `0xffff`, then here I think there will be no overflow.  The serialization for `KeyGroup` looks like this
```
	public static void writeKeyGroup(
		int keyGroup,
		int keyGroupPrefixBytes,
		DataOutputView keySerializationDateDataOutputView) throws IOException {
		for (int i = keyGroupPrefixBytes; --i >= 0; ) {
			keySerializationDateDataOutputView.writeByte(extractByteAtPosition(keyGroup, i));
		}
	}

	private static byte extractByteAtPosition(int value, int byteIdx) {
		return (byte) ((value >>> (byteIdx << 3)));
	}
```

Reply:
For max parallelism between `[1, 128]`, the key group range would be `[0, 127]`, which only has one byte for the serialization.
For max parallelism between `[129, 32767]`, the key group range would be `[128, 32766]` which could be interpreted as two bytes for serialization.

Reply:
Can we have a quick chat using DingDing?? I think there is some disagreement on the issue here.

Reply:
If the max parallelism is 128, the keyGroup-127 would be serialized as `0xff` with the single byte.

Reply:
The binary form of 127 should be `01111111`, not `11111111`.

Reply:
It's my bad to give the wrong description of the binary format of `127`.

You can run the code below:
~~~java
        int maxKeyGroup = 127;
        int keyGroupPrefixBytes =
                CompositeKeySerializationUtils.computeRequiredBytesInKeyGroupPrefix(
                        maxKeyGroup + 1);
        byte[] bytes = new byte[keyGroupPrefixBytes];
        bytes[0] = extractByteAtPosition(maxKeyGroup, 0);;
        byte[] upperBound = calculateUpperBound(bytes);
~~~
And you will get the byte array of `upperBound` as `[-128]`, from my mind,   changing `0x7f` to `0x80` for a single byte is a kind of **overflow**. However, this is legal in your code, which would make other developers confused.

Reply:
I ran this code and it returned `-128` as upperBound and no exception was thrown.

In my opinion, `RocksDB` just does seek matching by prefix, so here we should calculate as `unsigned` numbers, not as `signed` numbers.

In fact, we always treat the serialized result value as an `unsigned` number, not as a `signed` number. In [clipDBWithKeyGroupRange](https://github.com/apache/flink/blob/master/flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBIncrementalCheckpointUtils.java#L128), when `currentKeyGroupRange.getEndKeyGroup()` is 127, 
The result of `CompositeKeySerializationUtils.serializeKeyGroup(currentKeyGroupRange.getEndKeyGroup() + 1, endKeyGroupBytes)` is also -128. 

The overflow in this part of the code means that an overflow occurs in the calculation of `unsigned` numbers, which should not exist.

Reply:
Though RocksDB uses `unsigned` bytes, it will be really strange to treat `-128` with a single byte as not overflow on the java side.

As I said before, we will not meet the case to write `-128` within [clipDBWithKeyGroupRange](https://github.com/apache/flink/blob/master/flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBIncrementalCheckpointUtils.java#L128). During the restoring, we cannot change the max parallelism. If we can make the condition `currentKeyGroupRange.getEndKeyGroup() > targetKeyGroupRange.getEndKeyGroup()` pass, the max parallelism is larger than 128, which means the `prefixKeyGroupBytes` is 2. Thus, we will get `0x0080` for the prefix byte array of `127 + 1`.

Reply:
Assuming the max parallelism is 128(so we always serialize the KeyGroup with 1 byte.), the original parallelism is 1.  Then we get a DB instance of KeyGroupRange(0, 127), which is `currentKeyGroupRange`.

When our parallelism is adjusted to 2, our new `KeyGroup` assignments are [0, 63] and [64, 127]. For `KeyGroupRange`(0, 63) we delete data belonging to `KeyGroup`(64,127). Since the `deleteRange` of `RocksDB` contains the left border and does not contain the right border, in order to completely delete the data less than or equal to `KeyGroup`(127), we add 1 to the `KeyGroup`, which makes us actually serialize 128, and the serialized result is 0x80, also produces overflow in the Java sense. 

> CompositeKeySerializationUtils.serializeKeyGroup( **keyGroupRange.getEndKeyGroup() + 1**, stopKeyGroupPrefixBytes)

Reply:
If RocksDB uses unsigned numbers for matching, why do we need to compute with signed numbers here? If we handle overflow at the Java level by appending byte to the prefix, this will not get the correct result in RocksDB.

Reply:
I think you're right with the example of changing parallelism from 1 to 2 at the max parallelism of 128.

From my point of view, code blow looks easier to understand: the wrong case of increasing the byte array by 1 is that the current prefix byte array is `0xff` with a single byte or `0xffff` with two bytes:
~~~java
    protected final byte[] calculateUpperBound(byte[] prefix) {
        byte[] upperBound = new byte[prefix.length];
        System.arraycopy(prefix, 0, upperBound, 0, prefix.length);
        boolean overFlow = true;
        for (int i = prefix.length - 1; i >= 0; i--) {
            upperBound[i] = (byte) (prefix[i] + 1);
            if (prefix[i] != (byte) 0xff) {
                overFlow = false;
                break;
            }
        }
        Preconditions.checkArgument(!overFlow, "The upper boundary overflows.");
        return upperBound;
    }
~~~

Reply:
Resolved using the suggestion in the comment.